app/services/action_validator.rb

```diff
@@ -0,0 +1,39 @@
+class ActionValidator
+
+  def initialize(params)
+    @params = params
+  end
+
+  def form
+    @form ||= Form.includes(:form_elements).find(@params[:form_id])
+  end
+
+  def valid?
+    errors.size == 0
+  end
+
+  def errors
+    errors = []
+    form.form_elements.each do |element|
+      error = validate_field(element)
+      errors.push(error) unless error.nil?
+    end
+    errors
+  end
+
+  def validate_field(form_element)
+    key = form_element.name.to_sym
+    if form_element.required == true
+      if !@params.has_key?(key) || @params[key].blank?
+        return [form_element.name, "is required"]
+      end
+    end
+    if form_element.data_type == "email" && @params[key].present? && !is_email(@params[key])
```

Comment:
I like to wrap anything like this in a method with a descriptive name, or code readability:

```ruby
 if not_a_valid_email(form_element)
...

private
def not_a_valid_email(element)
  form_element.data_type == "email" && !is_email(@params[element.name.to_sym])
end